add in ACL extension support

urkle opened the pull request:
Requires djc/tokio-imap#142 (now merged, awaiting crate release) to handle the parsing-side of the responses.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jonhoo/rust-imap/227)
<!-- Reviewable:end -->

jonhoo commented on src/client.rs:
```diff
@@ -1280,6 +1280,91 @@ impl<T: Read + Write> Session<T> {
 
+    /// The [`SETACL` command](https://datatracker.ietf.org/doc/html/rfc4314#section-3.1)
+    /// Modifies the ACLs on the given mailbox for the specified identifier.
+    /// Return [`Error::No`] if the logged in user does not have `a` rights on the mailbox.
```
We should at the very least document that these methods aren't supported by all servers, and how users might check for server support. Alternatively, we could go one step further and do:

```rust
#[doc(alias = "RFC 4313")]
pub fn extension_acls(&mut self) -> Result<extensions::Acl<'_>> {
    if self.capabilities()?.has(Capability::TODO) {
        Ok(extensions::Acl(self))
    } else {
        Err("server does not support RFC 4314 ACL extensions")
    }
}
```

where all these methods are provided on the `extensions::Acl` type. That would also be a nice way to encapsulate other potential extensions in the future. What do you think?

jonhoo replied:
In fact, with GATs we could go even further and have

```rust
pub trait Extension {
    type Extended<'session>;
    fn extend<'session>(self, s: &'session mut Session) -> Result<Self::Extended<'session>>;
}

pub struct Rfc4314;

impl Extension for Rfc4314 {
    type Extended<'session> = extensions::Acl<'session>;
    fn extend<'session>(self, s: &'session mut Session) -> Result<Self::Extended<'session>> {
        if session.capabilities()?.has(Capability::TODO) {
            Ok(extensions::Acl(session))
        } else {
            Err("server does not support RFC 4314 ACL extensions")
        }
    }
}

impl Session {
    pub fn extension<E: Extension>(&mut self, ext: E) -> Result<E::Extended<'_>> {
        ext.extend(self)
    }
}
```

but given that GATs aren't stable I mostly just wanted to write that because it was interesting :)

urkle replied:
@jonhoo please look through the latest code changes and the two fixup commits that adjust the structure of things (the Cow'ifying).   Once we settle on those approaches then I'll work on this refactoring.

jonhoo replied:
I still think it'd be nice to do something like this, not just for this but also for the QRESYNC stuff that @bitfehler has been landing recently. We can do that in a separate PR though.

jonhoo replied:
Let's document that this (and the other ACL methods) only work for servers with the ACL-related capability listed though :+1:

jonhoo replied:
Leaving this unresolved to remember that we'll probably still want to make some change to how extensions are handled so that it's harder to misuse them and so that we don't pollute `Client` with tons and tons of methods.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,287 @@
+    }
+}
+
+impl ToString for AclRightList {
```
We should implement `Display` instead.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,287 @@
+    fn from(i: &str) -> Self {
+        i.chars()
+            .into_iter()
+            .map(|c| c.into())
```
This seems dubious. Is _every_ `char` a valid `AclRight`? What happens if one of the characters is some super obscure Unicode code point?

urkle replied:
the spec only allows lower alpha characters to be defined as extra rights. And the way I built the AclRight enum is that the core-defined rights (in 4314 and the annotate right in 5257) are enumerated values in the enum.  And any other "char" is put under a "Custom(char)" enumeration.  In this way any custom right provided by the server (via RIGHTS= capability) will be handled correctly by this iteration without issue. (I can add another test to show that case)

jonhoo replied:
Right, but this doesn't check for "lower alpha characters". It takes _any_ Unicode symbol (`char`). I think we'll want to have this fail if the `&str` contains any non-lower-ASCII-alpha characters.

urkle replied:
ok.. I'll change it to a TryFrom.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,287 @@
+#[derive(Debug, Eq, PartialEq)]
+pub struct Acl {
+    /// the mailbox these rights list are for
+    pub mailbox: String,
```
This should almost certainly be `Cow<'a, str>` so that we can do zero-copy parsing and not require copying to use this in a request.

urkle replied:
I've added a new commit that does this..  It makes things more complex IMHO.
See the changes here: d901b36273cff675a566c6b36e10070e56280112

jonhoo replied:
It does, but it also adds a feature :)

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,287 @@
+    }
+}
+
+impl Into<Acl> for imap_proto::types::Acl<'_> {
```
We should avoid this `impl` and make it an inherent method instead since it ends up being another public part of the API tied to `imap_proto`.

urkle replied:
@jonhoo is something like this what you are meaning here? d0a0f0d530

jonhoo replied:
Yep, exactly!

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,287 @@
+#[derive(Debug, Eq, PartialEq)]
+pub struct AclEntry {
+    /// The user identifier the rights are for
+    pub identifier: String,
```
`Cow<'a, str>`

jonhoo replied:
The same applies to the various later structs.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,287 @@
+    pub rights: AclRightList,
+}
+
+impl Into<AclEntry> for imap_proto::types::AclEntry<'_> {
```
Same here: inherent method.

jonhoo reviewed (changes requested):
Finally got around to taking a look at this. Thanks for doing all the groundwork! I left some notes inline.

urkle commented on Cargo.toml:
```diff
@@ -17,6 +17,9 @@ categories = ["email", "network-programming"]
 default = ["native-tls"]
 
+[patch.crates-io]
+imap-proto = { git = "https://github.com/djc/tokio-imap" }
```
I pinged the developer, asking about a release.

urkle replied:
They released a new version. so updated the cargo w/ the new dependency (0.16.1)

urkle commented on src/client.rs:
```diff
@@ -1280,6 +1280,91 @@ impl<T: Read + Write> Session<T> {
         .and_then(|lines| parse_id_seq(&lines, &mut self.unsolicited_responses_tx))
     }
 
+    /// The [`SETACL` command](https://datatracker.ietf.org/doc/html/rfc4314#section-3.1)
```
No problem..  It's your project, so nitpick away.

urkle commented on src/client.rs:
```diff
@@ -2020,6 +2104,146 @@ mod tests {
         assert_eq!(ids, [1, 2, 3, 4, 5].iter().cloned().collect::<Vec<_>>());
     }
 
+    #[test]
```
I can implement some integration tests,  The issue I see right now is I can build tests but only can test them on a real server (cyrus imap in my case). as Greenmail does not currently support the ACL extension. (they have comments in their code for where it would be added).   I've added a commit ( 85f8d0c6932cac714843fc5cce1b7da4df78b18a ) to pull out the host/port methods so that I can override ALL via ENV to allow me to more easily test that.  ( I can also provide a Dockerfile to setup a cyrus imap setup that will work like greenmail once I get the SMTP part done)

jonhoo replied:
Ahh, that's a shame. If you have a test that works locally against a real server, then I'm okay merging with that test not being run against Greenmail :+1: 

Having integration test setup against Cyrus as well as Greenmail would be amazing! The fact that we currently only test against a single server implementation is certainly a shortcoming.

urkle replied:
@jonhoo are you OK with skipping the integration tests in this PR and focusing on building those in #228 ?

jonhoo replied:
I think maybe we should try to land #228 so that we can then add the tests directly in this PR. What do you think?

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,351 @@
+#[self_referencing]
+pub struct Acl {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) acl: InnerAcl<'this>,
```
I think this would become nicer if you kept the mailbox in this struct. So:
```suggestion
    #[borrows(data)]
    #[covariant]
    pub(crate) mailbox: Cow<'this, str>,
    /// The list of identifier/rights pairs for the mailbox
    #[borrows(data)]
    #[covariant]
    pub(crate) acls: Vec<AclEntry<'a>>,
```

It'll also let you get rid of the `InnerAcl` type.

urkle replied:
I started that way, however the way ouroboros's build works it has separate parsers for each member. (mailbox_builder, and acls_builder both receicing the [u8] input). so I couldn't find any easy way to NOT end up parsing twice (which definitely would be undesirable). 

What I did instead was replace the named inner structs with inline anonymous tuples.

jonhoo replied:
Ahh, that's awkward. I think I'm in favor of reverting 00d3b288f2476b0dd87942815bbfb794cef43c00 then — tuple types with 4+ members get unwieldy and error-prone real fast!

urkle replied:
I reverted that commit.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,351 @@
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) rights: InnerListRights<'this>,
```
I don't think you need this inner type here either — you can just declare the fields directly here.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,351 @@
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) rights: InnerMyRights<'this>,
```
And same here.

jonhoo reviewed (changes requested):
Would you mind not force-pushing going forward — it makes reviewing incremental changes a pain (mostly because GitHub handles force-pushes so poorly) :sweat_smile:

jonhoo commented on src/types/acls.rs:
```diff
@@ -56,16 +56,32 @@ impl From<Vec<AclRight>> for AclRightList {
 
+/// Error from parsing AclRight strings
+#[derive(Debug, Eq, PartialEq)]
+pub enum AclRightError {
```
We'll definitely want to implement `std::error::Error` for this. I'm also not sure it needs to be an `enum`? If we truly expect new variants might be added in the future though, this should be marked `#[non_exhaustive]`.

urkle replied:
I added non_exhausted and implemented Error.  And by using an enum it makes it easier to extend in the future w/o breaking API.

jonhoo commented on src/client.rs:
```diff
@@ -1280,6 +1280,89 @@ impl<T: Read + Write> Session<T> {
+    /// Return [`Error::No`] if the logged in user does not have `a` rights on the mailbox.
+    pub fn set_acl<S1: AsRef<str>, S2: AsRef<str>>(
+        &mut self,
+        mailbox_name: S1,
+        identifier: S2,
```
nit: I'd probably just use `impl AsRef<str>` for each of these (and in the other methods).

urkle replied:
Now, FYI it this does follow the pattern you use for `fetch`,  `uid_fetch`, and `rename`.  but I can do that for these.

jonhoo replied:
Let's actually go ahead and change those too to be `impl AsRef<str>`. We should do so now since that's a breaking change.

urkle replied:
@jonhoo I added a new commit that changes all methods to use impl.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,398 @@
+use std::fmt::{write, Display, Formatter};
+use std::sync::mpsc;
+
+/// enum used for set_acl to specify how the ACL is to be modified.
```
We should use doc-links here. So `[Client::set_acl]` for example.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,398 @@
+
+/// Helpful wrapper around the ACL rights vector
+#[derive(Debug, Eq, PartialEq)]
+pub struct AclRightList {
```
Would this be better as just `AclRights`? Feels weird to call it a list when it's a set.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,398 @@
+#[derive(Debug, Eq, PartialEq)]
+pub struct InnerAcl<'a> {
+    /// The mailbox the ACL Entries belong to
+    pub(crate) mailbox: Cow<'a, str>,
```
I'm a little confused about why this and other fields are `Cow` and not just `&'a str` when it doesn't appear they're ever `Cow::Owned`, and will never be settable directly by the user?

urkle replied:
If I set them as `&'a str`, then the ouroboros builders fail to compile as the Cow is how they become referenced against the internal data struct. So there technically is an "owned" but it's just a transfer from the imap-proto type to my type (see line 127 in `Acl::parse`)

jonhoo replied:
Huh, that's very odd. I wonder if we could just match on the `Cow` to specifically grab the `'a` reference out of the `Cow::Borrowed` if we know it's always that variant. Or, phrased differently, I don't quite see how it'd ever end up being a `Cow::Owned` in that code, since the parser should already only be yielding things that borrow from `input` (I think). Unless it _sometimes_ does decoding I suppose, which could be the case.

urkle commented on src/types/acls.rs:
```diff
@@ -0,0 +1,399 @@
+                    match imap_proto::parser::parse_response(lines) {
+                        Ok((_rest, Response::Acl(a))) => {
+                            // lines = rest;
+                            return Ok(InnerAcl {
```
@jonhoo so, as I was implementing #235 I'm thinking this implementation for these methods is probably bad.  As it immediately returns out once it finds the Acl, which means it won't necessarily process all of the unsolicited responses.

jonhoo replied:
Oh, yes, good catch! It does indeed need to keep looping until it sees a `Done` or hits the end of the string I think, to catch cases where an unsolicited message appears between the `Acl` response and the `Done`.

urkle replied:
I have reworked the code now.

urkle commented:
@jonhoo let me know if things are good now w/ this PR..  and I'll rebase (resolving conflicts) and squash my commits down.

jonhoo commented on src/client.rs:
```diff
@@ -430,10 +430,10 @@ impl<T: Read + Write> Client<T> {
-        auth_type: S,
-        authenticator: &A,
+        auth_type: impl AsRef<str>,
+        authenticator: &impl Authenticator,
```
Ah, so, `Authenticator` I _think_ we may want to stay as `<A>`, since it's feasible that someone may want to specify the authenticator using turbofish syntax (which doesn't work with `:impl`). I don't know of a concrete case, I just have a feeling.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+use std::fmt::{Display, Formatter};
+use std::sync::mpsc;
+
+/// enum used for [Session::set_acl] to specify how the ACL is to be modified.
```
```suggestion
/// Specifies how [`Session::set_acl`] should modify an existing permission set.
```

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+    Remove,
+}
+
+/// Helpful wrapper around the ACL rights vector
+/// Used as input for [Session::set_acl] as output in [ListRights], [MyRights], and [AclEntry]
```
```suggestion
/// A set of [`AclRight`]s.
///
/// Used as input for [`Session::set_acl`] and as output in [`ListRights`], [`MyRights`], and [`AclEntry`].
```

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+
+impl AclRights {
+    /// Returns true if the AclRights has the provided ACL (either as a char or an AclRight enum)
+    pub fn has_right<T: Into<AclRight>>(&self, right: T) -> bool {
```
I wonder if we should just call this `has`?

jonhoo replied:
I think this is more akin to `HashSet` than `HashMap`. `HashMap::contains_key` has to specifically say "key" since there are also values, but `HashSet` just has `HashSet::contains`, so let's use that!

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+use std::sync::mpsc;
+
+/// enum used for [Session::set_acl] to specify how the ACL is to be modified.
+pub enum AclModifyMode {
```
This feels like it should be `Debug`, `Clone`, and `Copy` at the very least.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+            acl_builder: |input| {
+                let mut acls = Vec::new();
+
+                // There should only be ONE single ACL response
```
Let's also document why we still use `parse_many_input`.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+                if acls.is_empty() {
+                    Err(Error::Parse(ParseError::Invalid(input.to_vec())))
+                } else {
+                    Ok(acls.remove(0))
```
Should we error if there are >1?

urkle replied:
I've reworked this like I did in quota so it'll Err on != 1 (and adding tests)

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+
+/// Helpful wrapper around the ACL rights vector
+/// Used as input for [Session::set_acl] as output in [ListRights], [MyRights], and [AclEntry]
+#[derive(Debug, Eq, PartialEq)]
```
Any particular reason why this isn't also `Clone`?

urkle replied:
it wasn't needed by any of usage/tests :)  I'll add though.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The list of identifiers and rights for the [Acl] response
+#[derive(Debug, Eq, PartialEq)]
```
This also feels like it should be `Clone`.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+                if rights.is_empty() {
+                    Err(Error::Parse(ParseError::Invalid(input.to_vec())))
+                } else {
+                    Ok(rights.remove(0))
```
Same here — should we error if there are >1?

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+                if rights.is_empty() {
+                    Err(Error::Parse(ParseError::Invalid(input.to_vec())))
+                } else {
+                    Ok(rights.remove(0))
```
Same here — should we error if there are >1?

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,372 @@
+            rights_builder: |input| {
+                let mut rights = Vec::new();
+
+                // There should only be ONE single MYRIGHTS response
```
It does feel very weird to be using `parse_many_into` for this in so many places. I wonder if we should make a new helper, `parse_until_done`, that specifically handles the single-item case (and could internally call `parse_many_into` for now).

urkle replied:
I can look at that.

jonhoo reviewed (changes requested):
I did a pass over your new changes and a found a few more tweaks we should make.

jonhoo commented on src/parse.rs:
```diff
@@ -71,6 +71,31 @@ where
+/// If zero or more than one `T` is found then into will be `None`
+pub(crate) fn parse_until_done<'input, T, F>(
+    input: &'input [u8],
+    into: &mut Option<T>,
```
Instead of taking `&mut Option<T>`, let's just return a `T`

jonhoo commented on src/parse.rs:
```diff
@@ -71,6 +71,31 @@ where
+where
+    F: FnMut(Response<'input>) -> Result<MapOrNot<'input, T>>,
+{
+    let mut temp_output = Vec::<T>::new();
```
I think we _should_ be able to to have `parse_many_into` accept an `&mut Option` as `IT` directly. At least I thought `Option` implemented `Extend` as well.

jonhoo replied:
Huh, interesting. I could have sworn... Oh well, fair enough, let's do it this way then.

jonhoo commented on src/parse.rs:
```diff
@@ -71,6 +71,31 @@ where
+
+    *into = match temp_output.len() {
+        1 => Some(temp_output.remove(0)),
+        _ => None,
```
Any particular reason not to turn this into an error right here?

urkle replied:
So it leaves the decision up to the caller of whether it is an error (e.g. for cases where 0 or 1 response are acceptable).

jonhoo replied:
But they can't tell from this whether it's 0 or >1. I think it's okay to have it return an error on != 1, and then the caller can always recover the `Vec` from the error if they really want to.

urkle commented:
@jonhoo now that we have cyrus imapd testing in mainline I'll cherry pick those over (so as not to rebase yet) and add some integration tests in the next few days.

urkle commented:
@jonhoo I have pushed up some acl integration tests using a cargo feature to enable.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,350 @@
+///
+/// The list of identifiers and rights for the [`Acl`] response
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct AclEntry<'a> {
```
Let's set this to be `#[non_exhaustive]` in case other extensions we add support for in the future might add more fields.

jonhoo replied:
This still needs to be done.

urkle replied:
I added it to the wrong structs :)

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,350 @@
+    }
+
+    /// Returns the mailbox for the rights
+    pub fn mailbox(&self) -> &str {
```
Given the change we made around `Quota`, how do you feel about having just one accessor here that gives out `&InnerListRights` (and giving it a friendlier name)?

urkle replied:
If we went that route then it might make more sense to simply name the structs returned by the Session methods to be *Response (ListRightsResponse, MyRightsResponse) etc.. And then the "inner" structs will be renamed to ListRights, MyRights, etc..

Let me know what you think.

urkle replied:
@jonhoo I'll push a fixup commit that performs this change (*Response structs) so you can see it and then we can decide to keep or go another direction

jonhoo replied:
That seems reasonable to me!

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,350 @@
+
+/// Inner struct to manage storing the references for ouroboros
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) struct InnerListRights<'a> {
```
If we do make this `pub`, it should also be `#[non_exhaustive]`.

jonhoo commented on .github/workflows/test.yml:
```diff
@@ -32,3 +32,28 @@ jobs:
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
+          args: --features test-full-imap -- --ignored
```
With the new `--features` I don't think you also need `--ignored` any more?

jonhoo reviewed (commented):
I think you'll need to rebase this onto `master` so that GitHub will run CI again.

urkle commented:
@jonhoo I rebased + squashed.  then I added 2 fixups for the last set of changes you requested.

## 173add1
- remove test  --ignored
- add docs about server ACL capability is required
- non_exhaustive on AclRights
## be2c6a1
- rework Session method repones types to be *Response (MyRightsResponse AclResponse) and make public the now renamed "Inner*"

urkle commented:
@jonhoo all green! ✅

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,387 @@
+///
+/// Used as input for [`Session::set_acl`] as output in [`ListRights`], [`MyRights`], and [`AclEntry`]
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[non_exhaustive]
```
I don't think this needs `#[non_exhaustive]` since it has private fields already anyway.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,387 @@
+///
+/// Used by [`ListRightsResponse`].
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
```
This doesn't need to be `#[non_exhaustive]` since it has private fields. It would need `#[non_exhaustive]` if we made all those fields `pub` (and removed the accessor methods) though.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    /// Access to the wrapped [`ListRights`] struct
+    pub fn acl(&self) -> &Acl<'_> {
```
Instead of having this method be called something different for each type, how about a method by the same name in each one to access the "inner" type. Maybe it could be called `parsed`, `inner`, or `as_ref`?

urkle replied:
So, `inner` or `parsed` maybe as it is the object inside the wrap
`as_ref()` somewhat, but it is changing types which is not has as_ref() behaves on other rust structs with that contract.

Though once we get to the quota PR there is one of the Responses there that has two types it returns (a QuotaRoot and an array of Quotas)

In the end, your decision on this.

jonhoo replied:
I'm 50/50 on `inner` vs `parsed`. `inner` is more "standard", but `parsed` more accurately reflects what it is. I think either will be fine, so I'll leave it to you :)

For a type that has two "inners", I think it's totally fine for the accessors to be named.

urkle replied:
I pushed a fixup that uses parsed

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,387 @@
+///
+/// Used by [`MyRightsResponse`].
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
```
This also does not need to be `#[non_exhaustive]` if it's only going to have private fields.

jonhoo commented on src/client.rs:
```diff
@@ -443,9 +443,9 @@ impl<T: Read + Write> Client<T> {
+    fn do_auth_handshake(
         mut self,
-        authenticator: &A,
+        authenticator: &impl Authenticator,
```
This one too we should keep as `<A>`.

jonhoo commented on src/client.rs:
```diff
@@ -1282,14 +1282,112 @@ impl<T: Read + Write> Session<T> {
+    /// Modifies the ACLs on the given mailbox for the specified identifier.
+    /// Return [`Error::No`] if the logged in user does not have `a` rights on the mailbox.
+    ///
+    /// This method only works against a server with the ACL capability. Otherwise [`Error::Bad`]
```
Is "ACL" the actual name for the capability one should check for on the server? Or does it have some other name that'll be what has to be checked for?

urkle replied:
ACL is the actual name in the server.n (So `has_str("ACL")`)

There are other "sub-pieces" for determining additional functionality. e.g.
- checking `RIGHTS=kx`
   - which the imap-proto crate does not easily allow you to query as `RIGHTS=` lists extra rights characters supported by the server.  I'll have to do a new PR so support that one. (and make the Capability enum non_exhaustive)
- also the ANNOTATE-EXPERIMENT-1 capability to detect the annotation extension (adds the `n` ACL)

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,387 @@
+use std::fmt::{Display, Formatter};
+use std::sync::mpsc;
+
+/// Specifies how [`Session::set_acl`] should modify an existing permission set.
```
Clippy seems to think that this link doesn't work. Almost certainly because `Session` isn't in scope here. You can fix that by adding this to the `use` list:

```rust
#[cfg(doc)]
use crate::Session;
```

I think.

jonhoo commented on src/types/acls.rs:
```diff
@@ -0,0 +1,387 @@
+    InvalidRight,
+}
+
+impl Display for AclRightError {
```
Let's add a test for this `impl` if we're indeed going to ship it. I don't _think_ we actually rely on this `impl` any more, in which case it could arguably just be removed.

urkle replied:
Display (and Debug) are required in order to implement Error.

I can definitely add tests

jonhoo reviewed (changes requested):
Nice! Took another pass over the latest changes.

urkle commented:
@jonhoo there is an error in test caused by a new release of ouroboros joshua-maros/ouroboros#66  There is a pending PR to fix it, but has not yet been released.

jonhoo commented:
Sorry for the delay in review!

And yeah, I saw https://github.com/joshua-maros/ouroboros/issues/66

urkle commented:
@jonhoo pushed a fixup that renamed the "inner" methods to `parsed`.  Let me know if there is anything else.  If there is nothing let me know and I'll rebase + squash the commits.

jonhoo reviewed (approved):
Looks good to me!

jonhoo commented:
Released in alpha 9 :tada:
